### PR TITLE
8251358: Clean up Access configuration after Shenandoah barrier change

### PR DESCRIPTION
**Pull request by adityamandaleeka**
This change removes a bunch of code that is no longer necessary after the Shenandoah barrier changes, including the `SUPPORT` macros in barrierSetConfig.hpp and code using them. `Access::resolve` and associated code have also been removed, and callers have been modified.

Testing done:
tier1
hotspot_gc_shenandoah

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8251358](https://bugs.openjdk.java.net/browse/JDK-8251358): Clean up Access configuration after Shenandoah barrier change

### Reviewers
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**)
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to 45d775811f6befd497a0af4e1bc809930244bb69
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to 45d775811f6befd497a0af4e1bc809930244bb69

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/146/head:pull/146`
`$ git checkout pull/146`

### CODE REVIEW

**Comment by adityamandaleeka** [comment]
@kimbarrett Can you please take a look when you have time?

**Review by fisk — CHANGES_REQUESTED** [review]
Thanks for fixing this. It's nice to see build-time decorators go. Naturally, when removing the resolve() function, we can also remove a bunch of "_raw" functions in the runtime. Like there is no longer a need to distinguish between field_addr and field_addr_raw. Or between base and base_raw. They were all introduced so that one of them could apply the resolve barrier. The same goes for mark vs mark_raw and the other related mark accessor functions on oopDesc. I think we want to get rid of that too, right?

**Comment by adityamandaleeka** [comment]
@fisk Good point. It's clear that the field_addr and base ones can be changed, and I'll do that in the next update I send. I'll take a closer look at the marking functions you mentioned as well. I want to understand the decorators they're using and see whether they still do anything meaningful before I get rid of them.

**Comment by fisk** [comment]
The only supported decorators for primitives is memory ordering decorators. These functions use MO_RELAXED. Specifying MO_RELAXED is equivalent to using Atomic::load/store.

**Comment by adityamandaleeka** [comment]
@fisk Thanks again for the feedback, and sorry for the delay (it's been a very busy week!). I've pushed an update that cleans up a bunch of the `_raw` stuff.

**Review by rkennke — APPROVED** [review]
Looks good to me! Good to see those unused code paths go! Thanks!

**Comment by adityamandaleeka** [comment]
Thanks @rkennke . I will also need a sponsor for this change. Can you or @fisk please do that?

File: src/hotspot/share/gc/g1/g1FullGCCompactTask.cpp [hunk 1]
```diff
@@ -71,7 +71,7 @@ size_t G1FullGCCompactTask::G1CompactRegionClosure::apply(oop obj) {
   HeapWord* obj_addr = cast_from_oop<HeapWord*>(obj);
   assert(obj_addr != destination, "everything in this pass should be moving");
   Copy::aligned_conjoint_words(obj_addr, destination, size);
-  oop(destination)->init_mark_raw();
+  oop(destination)->init_mark();
```
Comment (shipilev) [inline src/hotspot/share/gc/g1/g1FullGCCompactTask.cpp:74]: For example, here.

File: src/hotspot/share/gc/g1/g1FullGCCompactTask.cpp [hunk 2]
```diff
@@ -48,7 +48,7 @@ class G1ResetHumongousClosure : public HeapRegionClosure {
         if (_bitmap->is_marked(obj)) {
           // Clear bitmap and fix mark word.
           _bitmap->clear(obj);
-          obj->init_mark_raw();
+          obj->init_mark();
```
Comment (shipilev) [inline src/hotspot/share/gc/g1/g1FullGCCompactTask.cpp:51]: For example, here.

File: src/hotspot/share/gc/parallel/psPromotionLAB.cpp [hunk 3]
```diff
@@ -83,7 +83,7 @@ void PSPromotionLAB::flush() {
   // so they can always fill with an array.
   HeapWord* tlab_end = end() + filler_header_size;
   typeArrayOop filler_oop = (typeArrayOop) top();
-  filler_oop->set_mark_raw(markWord::prototype());
+  filler_oop->set_mark(markWord::prototype());
```
Comment (shipilev) [inline src/hotspot/share/gc/parallel/psPromotionLAB.cpp:86]: ...or here.

**Review by shipilev — CHANGES_REQUESTED** [review]
No, wait a second. I agree with some `_raw` removals, but not the other. The point of many `_raw` calls is using them from internal GC code when we *know* barriers are not required (because GC already knows in which state the object is in) or even correct to perform (for example, setting marks on incomplete objects). Changing e.g. `set_mark_raw` to `set_mark` on those paths exposes objects to GC barriers.

**Comment by fisk** [comment]
> No, wait a second. I agree with some `_raw` removals, but not the other. The point of many `_raw` calls is using them from internal GC code when we *know* barriers are not required (because GC already knows in which state the object is in) or even correct to perform (for example, setting marks on incomplete objects). Changing e.g. `set_mark_raw` to `set_mark` on those paths exposes objects to GC barriers.

We did not have said raw functions before Shenandoah. They were added to distinguish between barriers vs no barriers. And people would have to figure out which one to call. Now we never use barriers, so there is no difference. Sure the GC code knows there are no barriers. Agreed. But so does everybody else. Because we are removing barriers for primitives support. And mark words are simply primitives. Or do you have any other need for actually having barriers on mark word reads and writes?

**Comment by shipilev** [comment]
> We did not have said raw functions before Shenandoah. 

Right. I misremembered the history here. JDK-8199735 added these `_raw` methods and this effectively reverts it.

**Review by shipilev — APPROVED** [review]

**Comment by shipilev** [comment]
Speaking of JDK-8199735, `cas_set_mark_raw` is also unnecessary now?

**Comment by fisk** [comment]
> Speaking of JDK-8199735, `cas_set_mark_raw` is also unnecessary now? 

Yes, good catch. I found 2 more mark word accessors that I commented on.

**Comment by adityamandaleeka** [comment]
Thanks @shipilev and @fisk . I had left `cas_set_mark_raw` alone since there's a caller that specifies `memory_order_release` and I didn't want to change any behavior with the cleanup: https://github.com/openjdk/jdk/blob/master/src/hotspot/share/gc/parallel/psPromotionManager.inline.hpp#L242

That said, I haven't looked into how necessary that is; perhaps it's okay to change it to relaxed order.

@fisk Thanks for taking another look. I don't see the new comments you added yet. Have you submitted them?

**Review by fisk — CHANGES_REQUESTED** [review]
Oops seems like I didn't publish the comments somehow.

**Comment by fisk** [comment]
> Thanks @shipilev and @fisk . I had left `cas_set_mark_raw` alone since there's a caller that specifies `memory_order_release` and I didn't want to change any behavior with the cleanup: https://github.com/openjdk/jdk/blob/master/src/hotspot/share/gc/parallel/psPromotionManager.inline.hpp#L242
> 
> 
> 
> That said, I haven't looked into how necessary that is; perhaps it's okay to change it to relaxed order.
> 
> 
> 
> @fisk Thanks for taking another look. I don't see the new comments you added yet. Have you submitted them?
> 
> 
> 
> I'll follow up on this PR tomorrow since it's quite late here 😄 
> 
> 

A lot of thought has gone into the memory ordering flags. This change should definitely not change any of that. Having said that, I can't see why we can not rename raw functions to not raw functions, while maintaining the same memory ordering.

**Comment by adityamandaleeka** [comment]
That makes sense, and I just pushed an update that renames those functions. `cas_set_mark_raw` had a default argument for the memory order which I've turned into a required argument in order to disambiguate after the rename. The only caller that was relying on the default argument was in the Shenandoah forwarding code, and I've updated it to pass in the memory order.

**Comment by adityamandaleeka** [comment]
@fisk Can you please take another look at this when you get a chance?

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by adityamandaleeka** [comment]
Thank you!

**Comment by adityamandaleeka** [comment]
Can one of the reviewers please sponsor the change?

**Pull request closed without merging**